File: Interview/src/array_string/Six.java
/*
 * Given an image represented by an NxN matrix, where each pixel in the image is 4 bytes, 
 * write a method to rotate the image by 90 degrees. Can you do this in place?
 */
package array_string;

/**
 *
 * @author Milad
 */
public class Six {
    public static void main(String[] args){
        char[][] matrix = {
            {'a','b','c'},
            {'d','e','f'},
            {'g','h','i'},
        };
        
        System.out.println(matrix.length);
        
        char[][] result = rotate(matrix);
        if(result == null) {
            System.out.println("Rows and Columns are not equal!");
        } else{
            for(int i = 0; i < result.length; i++){
                for(int j = 0; j < result.length; j++){
                    System.out.print(result[i][j] + "  ");
                }
                System.out.println("");
            }
        }
    }
    
    public static char[][] rotate(char[][] m1){
        
        int row = m1.length;
        int column = m1[0].length;
        if(row != column)
            return null;
        char[][] m2 = new char[row][column];
        
        for(int i = 0; i < row; i++){
            for(int j = 0; j < column; j++){
                m2[i][j] = m1[(row-1)-j][i];
            }
        }
        return m2;
    }
}

File: Python/Six.py
# Given an image represented by an NxN matrix, where each pixel in the image is 4 bytes,
# write a method to rotate the image by 90 degrees. Can you do this in place?

# -------------------------------------------------------

# Creates a list containing 5 lists, each of 8 items, all set to 0
# w, h = 8, 5;
# Matrix = [[0 for x in range(w)] for y in range(h)]


# https://docs.python.org/3/tutorial/datastructures.html#list-comprehensions
matrix = [[x**2 for x in range(3)] for y in range(3)]

print( matrix)

# print("matrix: " + len(matrix))
# print("rotated matrix: " + rotate(matrix))

def rotate(matrix):
    row = len(matrix)
    column = len(matrix[0])
    if row != column:
        return None
    m2 = [[0 for x in range(row)] for y in range(column)]
    print(m2)
    for i in range(row):
        for j in range(column):
            m2[i][j] = matrix[(row-1) - j][i]
    return m2

print(rotate(matrix))